linkerd/app/core/src/errors.rs

```diff
@@ -1,150 +1,177 @@
+    type Respond = Respond<B>;
+
+    fn new_respond(&self, req: &http::Request<A>) -> Self::Respond {
+        if req.version() == http::Version::HTTP_2 {
+            let is_grpc = req
+                .headers()
+                .get(http::header::CONTENT_TYPE)
+                .and_then(|v| v.to_str().ok().map(|s| s.starts_with("application/grpc")))
+                .unwrap_or(false);
+            Respond::Http2 { is_grpc }
+        } else {
+            Respond::Http1(self.0)
+        }
     }
 }
 
-impl<T, M> svc::Service<T> for Stack<M>
-where
-    M: svc::Service<T>,
-{
-    type Response = Service<M::Response>;
-    type Error = M::Error;
-    type Future = futures::future::Map<M::Future, fn(M::Response) -> Self::Response>;
-
-    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        self.inner.poll_ready()
-    }
-    fn call(&mut self, target: T) -> Self::Future {
-        self.inner.call(target).map(Service)
+impl<B> Clone for NewRespond<B> {
+    fn clone(&self) -> Self {
+        NewRespond(self.0)
     }
 }
 
-impl<S, B1, B2> svc::Service<Request<B1>> for Service<S>
-where
-    S: svc::Service<Request<B1>, Response = Response<B2>>,
-    S::Error: Into<Error>,
-    B2: Default,
-{
-    type Response = S::Response;
-    type Error = Error;
-    type Future = ResponseFuture<S::Future>;
-
-    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        self.0.poll_ready().map_err(Into::into)
-    }
+impl<B: Default> respond::Respond for Respond<B> {
+    type Response = http::Response<B>;
 
-    fn call(&mut self, req: Request<B1>) -> Self::Future {
-        let is_http2 = req.version() == Version::HTTP_2;
-        let inner = self.0.call(req);
-        ResponseFuture { inner, is_http2 }
-    }
-}
+    fn respond(&self, error: Error) -> Result<Self::Response, Error> {
+        tracing::warn!("Failed to proxy request: {}", error);
+
+        if let Respond::Http2 { is_grpc } = self {
+            if let Some(reset) = error.h2_reason() {
+                debug!(%reset, "Propagating HTTP2 reset");
+                return Err(error);
+            }
 
-impl<F, B> Future for ResponseFuture<F>
-where
-    F: Future<Item = Response<B>>,
-    F::Error: Into<Error>,
-    B: Default,
-{
-    type Item = Response<B>;
-    type Error = Error;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        match self.inner.poll() {
-            Ok(ok) => Ok(ok),
-            Err(err) => {
-                let err = err.into();
-
-                if self.is_http2 {
-                    if err.h2_reason().is_some() {
-                        debug!("propagating http2 response error: {:?}", err);
-                        return Err(err);
-                    }
-                }
-
-                let response = Response::builder()
-                    .status(map_err_to_5xx(err))
-                    .header(header::CONTENT_LENGTH, "0")
+            if *is_grpc {
+                debug!("Handling error with gRPC status code");
```

Comment:
should we move this to somewhere where we can record what status we're returning?